### ISSUE TITLE
Crash with Thai locale

### ISSUE THREAD

**Issue by lassana** [novotnyllc/MetroLog#110]
Both iOS and Android. It happens if Thai language is set on the device. 
Probably it's because of some Xamarin/mono bug ([#31228](https://bugzilla.xamarin.com/show_bug.cgi?id=31228), [#59077](https://bugzilla.xamarin.com/show_bug.cgi?id=59077)).

```
Exception java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1194)
 Caused by java.lang.reflect.InvocationTargetException:
 Caused by android.runtime.JavaProxyThrowable: System.NotSupportedException: Could not activate JNI Handle 0xbee3ea68 (key_handle 0x2d4d5153) of Java type 'md587fbd1eadef0ed8e203a8620aaff9dec/MainActivity' as managed type 'MyApplication.Droid.MainActivity'.
 ---> System.ArgumentOutOfRangeException: Not a valid calendar for the given culture. Parameter name: value
 at System.Globalization.DateTimeFormatInfo.set_Calendar (System.Globalization.Calendar value) [0x00154] in <4ef3d260c99b4993be4bb9e9e4e6ed03>:0
 at System.Globalization.DateTimeFormatInfo..ctor (System.Globalization.CultureData cultureData, System.Globalization.Calendar cal) [0x00022] in <4ef3d260c99b4993be4bb9e9e4e6ed03>:0
 at System.Globalization.CultureInfo.get_DateTimeFormat () [0x00033] in <4ef3d260c99b4993be4bb9e9e4e6ed03>:0
 at System.Globalization.CultureInfo.GetFormat (System.Type formatType) [0x0002f] in <4ef3d260c99b4993be4bb9e9e4e6ed03>:0
 at System.Globalization.DateTimeFormatInfo.get_CurrentInfo () [0x00021] in <4ef3d260c99b4993be4bb9e9e4e6ed03>:0
 at System.DateTimeOffset.ToString (System.String format) [0x00006] in <4ef3d260c99b4993be4bb9e9e4e6ed03>:0
 at MetroLog.InternalLogger.Log (MetroLog.LogLevel logLevel, System.String message, System.Exception ex) [0x0000d] in <fa81bd3494f14659846b8a7d2ba0d94b>:0
 at MetroLog.InternalLogger.Log (MetroLog.LogLevel logLevel, System.String message, System.Object[] ps) [0x00011] in <fa81bd3494f14659846b8a7d2ba0d94b>:0
 at MetroLog.InternalLogger.Info (System.String message, System.Object[] ps) [0x00000] in <fa81bd3494f14659846b8a7d2ba0d94b>:0
 at MetroLog.Internal.LogManager.GetLogger (System.String name, MetroLog.LoggingConfiguration config) [0x00049] in <fa81bd3494f14659846b8a7d2ba0d94b>:0
 at MetroLog.Internal.LogManager.GetLogger (System.Type type, MetroLog.LoggingConfiguration config) [0x0001b] in <fa81bd3494f14659846b8a7d2ba0d94b>:0
 at MetroLog.Internal.LogManager.GetLogger[T] (MetroLog.LoggingConfiguration config) [0x00000] in <fa81bd3494f14659846b8a7d2ba0d94b>:0
 at MyApplication.Droid.MainActivity..ctor () [0x00006] in <0ef92215c94d457a9ed002c3c35635b8>:0
 at (wrapper dynamic-method) System.Object:b237c4cd-e4b7-45ea-ac3c-c249f19e4dd3 (intptr,object[])
 at Java.Interop.TypeManager.n_Activate (System.IntPtr jnienv, System.IntPtr jclass, System.IntPtr typename_ptr, System.IntPtr signature_ptr, System.IntPtr jobject, System.IntPtr parameters_ptr) [0x000de] in <22e71d5347d44fc9b1e3f1fea379fb09>:0
 --- End of inner exception stack trace ---
 at Java.Interop.TypeManager.n_Activate (System.IntPtr jnienv, System.IntPtr jclass, System.IntPtr typename_ptr, System.IntPtr signature_ptr, System.IntPtr jobject, System.IntPtr parameters_ptr) [0x00167] in <22e71d5347d44fc9b1e3f1fea379fb09>:0
 at (wrapper dynamic-method) System.Object:4e2b0663-b11f-4bef-a195-e259f9cc06dc (intptr,intptr,intptr,intptr,intptr,intptr)
mono.android.TypeManager.n_activate (TypeManager.java)
mono.android.TypeManager.Activate (TypeManager.java:7)
md587fbd1eadef0ed8e203a8620aaff9dec.MainActivity.<init> (MainActivity.java:25)
java.lang.reflect.Constructor.newInstance (Constructor.java)
java.lang.Class.newInstance (Class.java:1650)
android.app.Instrumentation.newActivity (Instrumentation.java:1079)
android.app.ActivityThread.performLaunchActivity (ActivityThread.java:2640)
android.app.ActivityThread.handleLaunchActivity (ActivityThread.java:2873)
android.app.ActivityThread.access$900 (ActivityThread.java:181)
android.app.ActivityThread$H.handleMessage (ActivityThread.java:1482)
android.os.Handler.dispatchMessage (Handler.java:102)
android.os.Looper.loop (Looper.java:145)
android.app.ActivityThread.main (ActivityThread.java:6145)
java.lang.reflect.Method.invoke (Method.java)
java.lang.reflect.Method.invoke (Method.java:372)
com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:1399)
com.android.internal.os.ZygoteInit.main (ZygoteInit.java:1194)
```

**Comment by lassana** [comment]
I can confirm this is a Xamarin bug

**Comment by AntRemo** [comment]
@lassana I am running into this issue as well. 

How did you work around it?

Thanks

**Comment by lassana** [comment]
@AntRemo to be honest I don't quite remember. I know at first I have been setting the global `Culture` of the app to `en-US`, but my final solution was a custom MetroLog `Layout` with an explicit DateTime formatter, for instance "yyyy-MM-dd HH:mm:ss", actually it could be anything but "o".

**Comment by lassana** [comment]
@AntRemo sorry, I was wrong. The actual fix for the bug [#59077](https://bugzilla.xamarin.com/show_bug.cgi?id=59077) is to add the following code to your `LinkerPleaseInclude` file:

    public static void IncludeThaiLocale(System.Globalization.ThaiBuddhistCalendar include)
    {
        include = new System.Globalization.ThaiBuddhistCalendar();
    }

The Xamarin's linker removes this class so you have to include it manually.

**Comment by AntRemo** [comment]
@lassana thanks for the info. I was able to resolve it with the approach above. :)
